submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java

```diff
@@ -0,0 +1,104 @@
+import com.kyriba.submittal.service.SpeakerService;
+import lombok.Builder;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.SessionAttribute;
+
+import java.util.List;
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+
+/**
+ * @author M-ABL
+ */
+// TODO: setup security. Let's assume a filter retrieves a speaker's id by email/password and puts the id to the session.
+@RestController
+@RequestMapping(value = "/api/v1/speaker",
+    consumes = APPLICATION_JSON_UTF8_VALUE,
+    produces = APPLICATION_JSON_UTF8_VALUE)
+public class SpeakerController
+{
+  private final SpeakerService speakerService;
+
+
+  @Autowired
+  SpeakerController(final SpeakerService speakerService)
+  {
+    this.speakerService = speakerService;
+  }
+
+
+  @PostMapping(path = "/register")
+  public SpeakerResponse register(@RequestBody final SpeakerRequest speaker)
+  {
+    // TODO: can the controller invoke Notification MS directly in order to send a confirmation email?
+    String key = speakerService.register(
+        speaker.email, speaker.password, speaker.personalData, speaker.topics, speaker.conferenceId);
+    return SpeakerResponse.builder()
+        .email(speaker.email)
+        .confirmationKey(key)
+        .build();
+  }
+
+
+  @PostMapping(path = "/topics")
+  public TopicResponse proposeTopics(@SessionAttribute("speakerId") long speakerId,
+                                     @RequestBody final TopicRequest topicRequest)
+  {
+    SpeakerDataDto<List<TopicDto>> result = speakerService
+        .proposeTopics(speakerId, topicRequest.topics);
+    return TopicResponse.builder()
+        .speakerId(result.getSpeakerId())
+        .topics(result.getData())
+        .build();
+  }
+
+
+  @PutMapping(path = "/personal")
+  public void modifyPersonalData(@SessionAttribute("speakerId") long speakerId,
+                                 @RequestBody final PersonalInfoDto personalData)
+  {
+    speakerService.modifyPersonalData(speakerId, personalData);
+  }
+
+
+  @Builder
+  private static class TopicRequest
+  {
+    private List<TopicDto> topics;
+  }
+
+  @Builder
+  private static class TopicResponse
+  {
+    private long speakerId;
+    private List<TopicDto> topics;
+  }
+
+  @Builder
+  private static class SpeakerRequest
+  {
+    private String email;
+    private String password;
+    private PersonalInfoDto personalData;
+    private List<TopicDto> topics;
+    private long conferenceId;
+  }
+
+  @Builder
+  private static class SpeakerResponse
+  {
+    private String email;
+    private String confirmationKey;

```

Comment:
What is this?

Reply:
the user has to confirm the registration. He'll get an email with a key. With a specific link a new request to ConfirmationController is performed

Reply:
OK, I got the idea.